<launch>
   <!-- Determine this using rosrun pointgrey_camera_driver list_cameras.
       If not specified, defaults to first camera found. -->
  <arg name="left_camera_name" default="left" />
  <arg name="right_camera_name" default="right" />
  <arg name="left_camera_serial" default="31703351" />
  <arg name="right_camera_serial" default="32703551" />
  <arg name="calibrated" default="1" />
  <arg name="exposure_time" default="30000" />
  <arg name="rect_left" default="0" />
  <arg name="rect_top" default="0" />
  <arg name="rect_height" default="1024" />
  <arg name="rect_width" default="1280" />
  <arg name="image_data_format" default="XI_RGB24" />
  

   <!-- Choose visualization -->
   <arg name="rviz" default="true" />
   <arg name="rtabmapviz" default="false" />
   <arg name="local_bundle" default="true" />
   <arg name="stereo_sync" default="false" />

  <group ns="ximea_stereo">
    <node pkg="nodelet" type="nodelet" name="camera_nodelet_manager" args="manager" />

    <node pkg="nodelet" type="nodelet" name="left_camera_nodelet"
          args="load ximea_camera/XimeaCameraNodelet camera_nodelet_manager" >
      <param name="frame_id" value="stereo_camera" />
      <param name="serial" value="$(arg left_camera_serial)" />
      <param name="image_data_format" value="$(arg image_data_format)" />
      <param name="camera_name" value="$(arg left_camera_name)" />

      <!-- When unspecified, the driver will use the default framerate as given by the
           camera itself. Use this parameter to override that value for cameras capable of
           other framerates. -->
      <!-- <param name="frame_rate" value="15" /> -->

      <!-- Use the camera_calibration package to create this file -->
      <param name="camera_info_url" if="$(arg calibrated)"
             value="file://$(env HOME)/.ros/camera_info/$(arg left_camera_serial).yaml" />

      <param name="yaml_url" value="$(find ximea_camera)/config/ximeaLeft.yaml" />
    </node>    

    <node pkg="nodelet" type="nodelet" name="right_camera_nodelet"
          args="load ximea_camera/XimeaCameraNodelet camera_nodelet_manager" >
      <param name="frame_id" value="stereo_camera" />
      <param name="serial" value="$(arg right_camera_serial)" />
      <param name="image_data_format" value="$(arg image_data_format)" />
      <param name="camera_name" value="$(arg right_camera_name)" />

      <!-- When unspecified, the driver will use the default framerate as given by the
           camera itself. Use this parameter to override that value for cameras capable of
           other framerates. -->
      <!-- <param name="frame_rate" value="15" /> -->

      <!-- Use the camera_calibration package to create this file -->
      <param name="camera_info_url" if="$(arg calibrated)"
             value="file://$(env HOME)/.ros/camera_info/$(arg right_camera_serial).yaml" />

      <param name="yaml_url" value="$(find ximea_camera)/config/ximeaRight.yaml" />
    </node>    

    <node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc">
        <param name="approximate_sync" value="true" />
       <!-- <remap from="left" to="leftCamera" />
        <remap from="right" to="rightCamera" />  -->
    </node>

<!--
    <node pkg="nodelet" type="nodelet" name="image_proc_debayer"
          args="load image_proc/debayer camera_nodelet_manager">
    </node>
-->
  </group>

   <group ns="rtabmap">   
   
      <!-- Stereo Odometry -->   
      <node pkg="rtabmap_ros" type="stereo_odometry" name="stereo_odometry" output="screen">
         <remap from="left/image_rect"       to="/ximea_stereo/left/image_rect"/>
         <remap from="right/image_rect"      to="/ximea_stereo/right/image_rect"/>
         <remap from="left/camera_info"      to="/ximea_stereo/left/camera_info"/>
         <remap from="right/camera_info"     to="/ximea_stereo/right/camera_info"/>         

         <param name="approx_sync"  type="bool" value="true"/>

         <param name="subscribe_rgbd"  type="bool" value="false"/>
         <param name="frame_id"        type="string" value="base_link"/>
         <param name="odom_frame_id"   type="string" value="odom"/>

         <param name="Odom/Strategy"      type="string" value="0"/> <!-- 0=Frame-to-Map, 1=Frame=to=Frame -->
         <param name="Vis/EstimationType" type="string" value="1"/> <!-- 0=3D->3D 1=3D->2D (PnP) -->
         <param name="Vis/MaxDepth"       type="string" value="0"/>
         <param name="Odom/GuessMotion" type="string" value="true"/>
         <param name="Vis/MinInliers"     type="string" value="10"/>
         <param unless="$(arg local_bundle)" name="OdomF2M/BundleAdjustment" type="string" value="0"/>
         <param name="OdomF2M/MaxSize"    type="string" value="1000"/> 
         <param name="GFTT/MinDistance"   type="string" value="10"/>
         <param name="GFTT/QualityLevel"  type="string" value="0.00001"/> 
         <param name="GFTT/QualityLevel"  type="string" value="0.00001"/>
         <param name="Stereo/MaxDisparity" type="int" value="128" />
         <param name="Stereo/OpticalFlow" type="bool" value="false" />
      </node>
   </group>      
   <param name="robot_description" textfile="$(find autorally_core)/launch/robot_stereo.urdf"/>

   <node pkg="robot_state_publisher_urdf" type="robot_state_publisher_urdf" name="robot_state_pub">
        <param name="publish_freq" type="int" value="20" />
   </node>

</launch>
